functions.h

```diff
@@ -105,6 +105,8 @@ template<typename idx_t>
 struct Indexer{
     idx_t idx;
     Indexer(): idx(0){}
+    template<class T>
+    Indexer(T obj): idx(obj->idx){}
```

Comment:
Shouldn't we pass an idx more than an obj?
What is done seems clearer with an idx, no?

Reply:
```
Indexer(T obj): idx(obj->idx){}
```
it's a typed interface, which means you cannot pass whatever you want. :)